src/System.Text.Json/src/System/Text/Json/JsonConstants.cs

```diff
@@ -63,7 +63,7 @@ internal static class JsonConstants
 
         public const int MaxEscapedTokenSize = 1_000_000_000;   // Max size for already escaped value.
         public const int MaxUnescapedTokenSize = MaxEscapedTokenSize / MaxExpansionFactorWhileEscaping;  // 166_666_666 bytes
-        public const int MaxBase46ValueTokenSize = (MaxEscapedTokenSize >> 2 * 3) / MaxExpansionFactorWhileEscaping;  // 125_000_000 bytes
+        public const int MaxBase64ValueTokenSize = (MaxEscapedTokenSize >> 2) * 3 / MaxExpansionFactorWhileEscaping;  // 125_000_000 bytes
```

Comment:
What forces us to have a hardcoded maximum?  Why not just let the implementation OOM if the required size would be too big?

Reply:
It may not OOM, it may cause integer overflow depending on the size when we calculate the maximum required size for the given input. And then, we may pass in an unexpected value to `IBW.GetSpan(...)`. This could be a user-defined IBW which could have custom logic for how the memory is allocated.

https://github.com/dotnet/corefx/blob/a6e7ffd72db078a46612afa9d085bd27ef8e2fed/src/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Bytes.cs#L50-L63

https://github.com/dotnet/corefx/blob/a6e7ffd72db078a46612afa9d085bd27ef8e2fed/src/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.cs#L1047-L1055

For example. if this calculation overflows (which currently is guaranteed not to), then we skip the call to Grow and then overwrite existing data, or only write partial data, or throw ArgumentException when trying to copy to a buffer that's too small:
https://github.com/dotnet/corefx/blob/a6e7ffd72db078a46612afa9d085bd27ef8e2fed/src/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Bytes.cs#L90-L95

Reply:
> it may cause integer overflow

And you can't use `checked`?

Reply:
Hmm, let me see if that is feasible, and follow up on this as a separate PR.

This PR is intended to fix the bug in the current code that was blocking user scenario (considering for 3.0), so I'd like to keep the change isolated to just that.

Reply:
> And you can't use `checked`?

Aren't their performance implications of doing so?

Reply:
> Aren't their performance implications of doing so? Or are you suggesting those would be negligible here?

It would add a few instructions, but you'd also get to remove calls like `JsonWriterHelper.ValidateBytes(bytes);`, simplify the code, and avoid these strange hard-coded, semi-accurate limits that appear to come out of nowhere.

Reply:
Good point. I am trying to see the side effects of removing these checks. Currently, the error is detected up-front with a deterministic error message. Removing the checks and letting integer overflow/OOM/etc. capture the failure condition based on the input and method being called may not be worth the trade-off. It is easy to reason about these limits today.

For example, let's say we remove the escaping/transcoding based limits. Today, we have the following up-front check on the public API:
https://github.com/dotnet/corefx/blob/06007cdedee9cfb096d884016a99e3a64bdccb45/src/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteProperties.String.cs#L454-L456
https://github.com/dotnet/corefx/blob/06007cdedee9cfb096d884016a99e3a64bdccb45/src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.cs#L117-L121

After that, all subsequent method calls can rely on this invariant.

Now, we remove that, and then the "error" state (i.e. data too large) propagates to all the places where the previous assertions break:
- GetMaxEscapedLength would need to be checked.
https://github.com/dotnet/corefx/blob/06007cdedee9cfb096d884016a99e3a64bdccb45/src/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteProperties.String.cs#L1010
- Calculating the required max size would need to be checked for both indented and minimized.
https://github.com/dotnet/corefx/blob/06007cdedee9cfb096d884016a99e3a64bdccb45/src/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteProperties.String.cs#L1477-L1482
https://github.com/dotnet/corefx/blob/06007cdedee9cfb096d884016a99e3a64bdccb45/src/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteProperties.String.cs#L1329-L1336

And now the user gets IntegerOverflowException rather than an ArgumentException.

That said, I think it's worth doing the exercise and see what code paths would need to be fixed and if reasoning about the kinds of errors that might get surfaced to the caller is easy enough to justify the change.

Reply:
`checked` would require catching the overflow exception. It's a forbidden exception type that should not bubble out to callers.

Reply:
> It's a forbidden exception type that should not bubble out to callers.

Is it forbidden by a known good practice, that other parts of framework are lacking? For example, on macOS we get this exception logged:

> System.OutOfMemoryException: Insufficient memory to continue the execution of the program.

with `System.IO.File.ReadAllText(pathTo3gbFile);`.

Reply:
He was referring to the overflow exception, not the out of memory exception.

Reply:
Sorry, i misread it (tired eyes). Just interested if there are good practices around such system exceptions and what is ok for APIs to passthru and when must they be handled.

Reply:
> This PR is intended to fix the bug in the current code that was blocking user scenario (considering for 3.0), so I'd like to keep the change isolated to just that.

Yes we need to keep this as simple as possible for 3.0 to avoid any potential regression. For 5.0 then we improve this area by removing the constants and use `checked` (if that turns out to be the best design).